Sources/NIO/SelectorUring.swift

```diff
@@ -0,0 +1,353 @@
+            filter |= Uring.POLLIN
+        }
+        if uringFilters.contains(.output) {
+            filter |= Uring.POLLOUT
+        }
+        if uringFilters.contains(.readHangup) {
+            filter |= Uring.POLLRDHUP
+        }
+        assert(filter & Uring.POLLHUP != 0) // both of these are reported
+        assert(filter & Uring.POLLERR != 0) // always and can't be masked.
+        return filter
+    }
+
+    fileprivate init(uringEvent: UInt32) {
+        var selectorEventSet: SelectorEventSet = ._none
+        if uringEvent & Uring.POLLIN != 0 {
+            selectorEventSet.formUnion(.read)
+        }
+        if uringEvent & Uring.POLLOUT != 0 {
+            selectorEventSet.formUnion(.write)
+        }
+        if uringEvent & Uring.POLLRDHUP != 0 {
+            selectorEventSet.formUnion(.readEOF)
+        }
+        if uringEvent & Uring.POLLHUP != 0 || uringEvent & Uring.POLLERR != 0 {
+            selectorEventSet.formUnion(.reset)
+        }
+        self = selectorEventSet
+    }
+}
+
+extension Selector: _SelectorBackendProtocol {
+    internal func _debugPrint(_ s : @autoclosure () -> String) {
+        #if SWIFTNIO_LIBURING_DEBUG_SELECTOR
+        print("S [\(NIOThread.current)] " + s())
+        #endif
+    }
+
+    func initialiseState0() throws {
+        // fail using uring unless it was successfully initialized
+        if Uring.initializedUring == false {
+            throw UringError.loadFailure
+        }
+
+        try ring.io_uring_queue_init()
+        self.selectorFD = ring.fd()
+
+        // eventfd are always singleshot and re-register each time around
+        // as certain use cases of nio seems to generate tons of wakeups
+        // (at least its tested for that in some of the performance tests
+        // e.g. future_whenallsucceed_100k_deferred_off_loop, future_whenallcomplete_100k_deferred_off_loop
+        // ) - if using normal ET multishots, we would get 100k events to handle basically.
+        // so using single shot for wakeups makes those tests run 30-35% faster approx.
+
+        self.eventFD = try EventFd.eventfd(initval: 0, flags: Int32(EventFd.EFD_CLOEXEC | EventFd.EFD_NONBLOCK))
+
+        ring.io_uring_prep_poll_add(fd: Int32(self.eventFD), pollMask: Uring.POLLIN, sequenceIdentifier:0, multishot:false) // wakeups
+
+        self.lifecycleState = .open
+        _debugPrint("UringSelector up and running fd [\(self.selectorFD)] wakeups on event_fd [\(self.eventFD)]")
+    }
+
+    func deinitAssertions0() {
+        assert(self.eventFD == -1, "self.eventFD == \(self.eventFD) on deinitAssertions0 deinit, forgot close?")
+    }
+
+    func register0<S: Selectable>(selectable: S, fd: Int, interested: SelectorEventSet, sequenceIdentifier: RegistrationSequenceIdentifier) throws {
+        _debugPrint("register interested \(interested) uringEventSet [\(interested.uringEventSet)] sequenceIdentifier[\(sequenceIdentifier)]")
+        deferredReregistrationsPending = true
```

Comment:
do we still need this? If yes, it's missing a `self.`

Reply:
Yes, related to comment above about deferring actual flushing of registration modifications until entering `whenReady0`. Could probably discuss if this would be useful behaviour overall for other selector backends (for example, for `kqueue`, this could decrease the number of syscalls significantly if multiple registration changes are done during a single event loop tick, as all of them would be coalesced/postpostponed until entering `whenReady0` the next time).

Reply:
Added self.